compiler/rustc_mir_transform/src/match_branches.rs

```diff
@@ -399,7 +401,10 @@ impl<'tcx> SimplifyMatch<'tcx> for SimplifyToExp {
                             if ((f_c.const_.ty().is_signed() || discr_ty.is_signed())
```

Comment:
```suggestion
                            if ((f_c.const_.ty().is_signed() && discr_ty.is_signed())
```

Reply:
Honestly I think it's best to move this entire thing into a helper, not just `int_equal`.

And also, why is `int_equal` so different from what happens for unsigned? They should be entirely identical except for using `int` vs `uint` functions.

Reply:
Ah I see, it's about converting the `u128` from the `SwitchInt` to a `ScalarInt`. But that should be uniform -- do it once before the comparison.  The input is not sign extended so it's always `SwitchInt::try_from_uint`.

So already further up, you should convert `first_val` and `second_val` to `ScalarInt`.

In fact we should probably change SwitchInt to store a `ScalarInt` rather than a raw `u128`.

Reply:
Hold on a sec... isn't what you actually want to do here some sort of cast? I don't know the right direction, but -- basically you want to cast the discriminant value to the type of the constant (or the other way around), and then check they are equal, right?

The interpreter has the `int_to_int_or_float` method for that.  But really you just care about [this match arm](https://github.com/rust-lang/rust/blob/989660c3e6efc1c5eb2e822f68863df7d06cbcb4/compiler/rustc_const_eval/src/interpret/cast.rs#L279-L288). That should probably be turned into a helper somewhere it can be used by mir-opts.

Reply:
> Otherwise you're still potentially treating something as signed that is unsigned, or vice versa.

Everything looks fine here. Any signed integer will be converted to a signed comparison.

Reply:
> In fact we should probably change SwitchInt to store a `ScalarInt` rather than a raw `u128`, or at least make it easy to get the SwitchTarget values as `ScalarInt`.

I have seen your new issue. :)

Reply:
> The interpreter has the `int_to_int_or_float` method for that. But really you just care about [this match arm](https://github.com/rust-lang/rust/blob/989660c3e6efc1c5eb2e822f68863df7d06cbcb4/compiler/rustc_const_eval/src/interpret/cast.rs#L279-L288). That should probably be turned into a helper somewhere it can be used by mir-opts.

Perhaps this could be a separate PR?

Reply:
> Everything looks fine here. Any signed integer will be converted to a signed comparison.

You're treating *both* values as signed if *either* of them is signed. That means you can be treating unsigned values as signed, which is wrong.

> Perhaps this could be a separate PR?

Perhaps, but I don't understand you current PR, so it may also be a way to turn this code into something that makes sense to more than one person. ;)

Reply:
> > Everything looks fine here. Any signed integer will be converted to a signed comparison.
> 
> You're treating _both_ values as signed if _either_ of them is signed. That means you can be treating unsigned values as signed, which is wrong.

In the known test cases, it is correct. But I must carefully check the edge cases here. For safety reasons, I will later consider only signed-to-signed conversions in this PR. :)

cc @rust-lang/wg-mir-opt Perhaps someone else will directly point out this specific problem?

Reply:
> In the known test cases, it is correct.

That's a very low bar. The comments should give convincing reasons why it is correct *for all possible MIR ever*.